File: src/main/java/xyz/upperlevel/hgame/world/scheduler/Task.java
package xyz.upperlevel.hgame.world.scheduler;

import lombok.Getter;

public abstract class Task implements Runnable {
    @Getter
    private int id;

    public Task() {
    }

    private boolean tryCancel() {
        if (id != -1) {
            Scheduler.cancel(id);
            id = -1;
            return true;
        }
        return false;
    }

    public void start(long delay, boolean repeat) {
        tryCancel();
        id = Scheduler.start(this, delay, repeat);
    }

    public void delay(long delay) {
        start(delay, false);
    }

    public void repeat(long each) {
        start(each, true);
    }

    public boolean isCanceled() {
        return id == -1;
    }

    public void cancel() {
        if (!tryCancel()) {
            throw new IllegalStateException("This task has already been cancelled.");
        }
    }
}

File: src/main/java/xyz/upperlevel/hgame/world/scheduler/Task.kt
package xyz.upperlevel.hgame.world.scheduler

abstract class Task : Runnable {
    var id: Int = 0
        private set

    val isCanceled: Boolean
        get() = id == -1

    private fun tryCancel(): Boolean {
        if (id != -1) {
            Scheduler.cancel(id)
            id = -1
            return true
        }
        return false
    }

    fun start(delay: Long, repeat: Boolean) {
        tryCancel()
        id = Scheduler.start(this, delay, repeat)
    }

    fun delay(delay: Long) {
        start(delay, false)
    }

    fun repeat(each: Long) {
        start(each, true)
    }

    fun cancel() {
        if (!tryCancel()) {
            throw IllegalStateException("This task has already been cancelled.")
        }
    }
}
